framework/src/kernels/NonlocalKernel.C

```diff
@@ -0,0 +1,101 @@
+/****************************************************************/
+/*               DO NOT MODIFY THIS HEADER                      */
+/* MOOSE - Multiphysics Object Oriented Simulation Environment  */
+/*                                                              */
+/*           (c) 2010 Battelle Energy Alliance, LLC             */
+/*                   ALL RIGHTS RESERVED                        */
+/*                                                              */
+/*          Prepared by Battelle Energy Alliance, LLC           */
+/*            Under Contract No. DE-AC07-05ID14517              */
+/*            With the U. S. Department of Energy               */
+/*                                                              */
+/*            See COPYRIGHT for full restrictions               */
+/****************************************************************/
+
+#include "NonlocalKernel.h"
+#include "Assembly.h"
+#include "MooseVariable.h"
+#include "Problem.h"
+#include "SubProblem.h"
+#include "SystemBase.h"
+
+// libmesh includes
+#include "libmesh/threads.h"
+#include "libmesh/quadrature.h"
+
+template<>
+InputParameters validParams<NonlocalKernel>()
+{
+  InputParameters params = validParams<Kernel>();
+  return params;
+}
+
+NonlocalKernel::NonlocalKernel(const InputParameters & parameters) :
+    Kernel(parameters)
+{
+}
+
+void
+NonlocalKernel::computeJacobian()
+{
+  Kernel::computeJacobian();
+
+  DenseMatrix<Number> & keg = _assembly.jacobianBlockNonlocal(_var.number(), _var.number());
+  _nonlocal_ke.resize(keg.m(), keg.n());
+  _nonlocal_ke.zero();
+
+  // compiling set of global IDs for the local DOFs on the element
+  std::set<dof_id_type> local_dof_indices;
+  for (_j = 0; _j < _phi.size(); _j++)
+    local_dof_indices.insert(_var.dofIndices()[_j]);
+
+  for (_i = 0; _i < _test.size(); _i++)
+    for (_k = 0; _k < _var.allDofIndices().size(); _k++)
```

Comment:
Ahhh!

What is this!

You're calling `allDofIndices()` every time through the loop!  That's doing a full mesh sweep every time through the loop!  Pull this out before you start looping... and save off the size!